File: Solutions/6kyu/6kyu_sum_of_threes.js
function sumOfThrees(n) {
  var powers = []
  
  while (n > 0) {
    if (n == 2) { return "Impossible" }
    var pow = getPow(n)
    powers.push(pow); n -= Math.pow(3,pow)
  }
  
  return (new Set(powers)).size == powers.length ?
            powers.map(e=> '3^'+e).join('+')
            :
            "Impossible"
  
}


function getPow(lim, p=0) {
  while (true) {
    if ( Math.pow(3,p) > lim) {
      return p-1
    }
    p++
  }   
}

File: Solutions/6kyu/6kyu_sum_of_threes.py
def sum_of_threes(n):
    
    powers = []

    while n > 0:
        if n == 2: return 'Impossible'
        res = gen_power(n); n -= pow(3, res)
        powers.append(res)

    return '+'.join('3^'+str(p) for p in powers) \
               if len(set(powers)) == len(powers) else 'Impossible'
    
    
def gen_power(lim, p=0):
    while True:
        if pow(3,p) > lim:
            return p-1
        p += 1
